File: 040-combination-sum-ii.cc
#include "leetcode.hh"

class Solution {
public:
    void dfs(vector<vector<int>>& results, vector<int>& progress,
             vector<int>& candidates, int start, int target, map<int, int>& counter) {
        if (target == 0) {
            results.emplace_back(progress);
            return;
        }

        for (int i = start; i < candidates.size(); i++) {
            int curr = candidates[i];
            if (curr > target) {
                break;
            }

            for (int j = 1; j <= counter[curr]; j++) {
                for (int k = 0; k < j; k++) {
                    progress.push_back(curr);
                }
                dfs(results, progress, candidates, i+1, target-curr*j, counter);
                for (int k = 0; k < j; k++) {
                    progress.pop_back();
                }
            }
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> progress;
        vector<vector<int>> result;
        map<int, int> counter;

        sort(candidates.begin(), candidates.end());

        for (int i = 0; i < candidates.size(); i++) {
            if (counter.find(candidates[i]) == counter.end()) {
                counter[candidates[i]] = 1;
            } else {
                counter[candidates[i]]++;
            }
        }

        candidates.erase(unique(candidates.begin(), candidates.end()), candidates.end());

        dfs(result, progress, candidates, 0, target, counter);
        return result;
    }
};


int main() {
    auto s = Solution();
    // 1,1,6
    vector<int> vec {10,1,2,7,6,1,5};
    auto result = s.combinationSum2(vec, 8);
    print_2d_vec(result);
    return 0;
}

File: 040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, arr: List[int], target: int) -> List[List[int]]:
        # dfs backtrack
        arr.sort()
        results = []
        n = len(arr)

        def traverse(i, target, path):
            if target == 0:
                results.append(path[:])
            for j in range(i, n):
                if arr[j] > target:
                    break
                if j > i and arr[j] == arr[j-1]:
                    continue
                path.append(arr[j])
                traverse(j+1, target-arr[j], path)
                path.pop()

        traverse(0, target, [])
        return results
